src/parser.jl

```diff
@@ -0,0 +1,246 @@
+using Base.Meta: isexpr
+
+
+"""
+A parser for a simple human-readable of an MOI instance.
+This should be thought of as a compact way to write small instances
+for tests, and not an exchange format.
+
+variables: x, y, z
+minobjective: 2x + 3y
+con1: x + y <= 1
+con2: [x,y] in Set
+
+special labels: variables, minobjective, maxobjective
+everything else denotes a constraint with a name
+all constraints must be named
+"x - y" does NOT currently parse, needs to be written as "x + -1.0*y"
+"x^2" does NOT currently parse, needs to be written as "x*x"
+"""
+
+struct ParsedScalarAffineFunction
```

Comment:
Does it make sense to change [`MOI.ScalarAffineFunction{T}`](https://github.com/JuliaOpt/MathOptInterface.jl/blob/master/src/functions.jl#L54) to
```julia
struct ScalarAffineFunction{V,T}
    variables::Vector{V}
    coefficients::Vector{T}
    constant::T
end
```
to avoid duplicating all these types?

Reply:
I'm a bit worried about this imposing extra cognitive load on everyone using MOI just to save some code in this case. But if @blegat also gives a thumbs up here then I'll go with it.

Reply:
Having thought about this a bit more, I agree with Miles. Best to keep MOI simple with only one concept for a variable (VariableReference).

Reply:
If we find two names we could do
```julia
struct Name1{V,T}
    variables::Vector{V}
    coefficients::Vector{T}
    constant::T
end
const Name2 = Name1{VariableReference}
```
And if `Name2 = ScalarAffineFunction` we do not even have to change any code.
For `Name1` we could take `ScalarAffineExpression`.

Reply:
or `GenericScalarAffineFunction` and `ScalarAffineFunction` to be consistent with JuMP's `GenericAffExpr` and `AffExpr`.

Reply:
Yes, that is a better idea. It will be in fact exactly the same definition as for the JuMP scalar type :-P